main.go

```diff
@@ -335,15 +336,42 @@ func prepareConfig(parseOpts *cliOptions, conf *ffuf.Config) error {
 		}
 	}
 
-	//Prepare headers
+	//Prepare headers and make canonical
 	for _, v := range parseOpts.headers {
 		hs := strings.SplitN(v, ":", 2)
 		if len(hs) == 2 {
-			conf.Headers[strings.TrimSpace(hs[0])] = strings.TrimSpace(hs[1])
+			// trim and make canonical
+			// except if used in custom defined header
+			var CanonicalNeeded bool = true
+			for _, a := range conf.CommandKeywords {
+				if a == hs[0] {
+					CanonicalNeeded = false
+				}
+			}
+			// check if part of InputProviders
+			if CanonicalNeeded {
+				for _, b := range conf.InputProviders {
+					if b.Keyword == hs[0] {
+						CanonicalNeeded = false
+					}
+				}
+			}
+			if CanonicalNeeded {
+				var CanonicalHeader string = textproto.CanonicalMIMEHeaderKey(strings.TrimSpace(hs[0]))
+				conf.Headers[CanonicalHeader] = strings.TrimSpace(hs[1])
+			} else {
+				conf.Headers[strings.TrimSpace(hs[0])] = strings.TrimSpace(hs[1])
+			}
 		} else {
 			errs.Add(fmt.Errorf("Header defined by -H needs to have a value. \":\" should be used as a separator"))
 		}
 	}
+
+	// set default User-Agent header if not present
+	if _, ok := conf.Headers["User-Agent"]; !ok {
+		conf.Headers["User-Agent"] = fmt.Sprintf("%s v%s", "Fuzz Faster U Fool", ffuf.VERSION)
+	}
+
```

Comment:
Unfortunately I think we need to move this part of the logic back to the `simple.go`:

If one of the headers is `-H "FUZZ: somevalue"`, this gets added to the list of headers as there's no header named `User-Agent` present, only the keyword. In go, maps are not ordered so there's a considerable chance that `FUZZ: somevalue` gets replaced and added to the result map _before_ we iterate over this native `User-Agent`. The `User-Agent: Fuzz Faster U Fool ...` will then replace the `User-Agent: somevalue` we added to the result map before.

If we do conditional statements in the `main.go` to figure out if we should add it in the first place, we still would need to do some checks and additions in `simple.go`. Because of this, I think the right call is to move this back to `simple.go`, after the canonicalization and replacing the keyword values. It should work as it is though.

Reply:
No problem. This is still far less complex than the Go language client for Ethereum. ;-)
I will move the codeblock back to simple.go.

Maybe another PR could be to build unit-test cases for all possible combinations.